public/app/features/explore/Explore.tsx

```diff
@@ -133,31 +140,114 @@ interface ExploreProps {
 
   constructor(props: ExploreProps) {
     super(props);
+
     this.exploreEvents = new Emitter();
   }
 
-  componentDidMount() {
+  stateSelector(state: StoreState): ReduxState {
+    const explore = state.explore;
+    const { split, syncedTimes } = explore;
+    const item: ExploreItemState = explore[this.props.exploreId];
+    const timeZone = getTimeZone(state.user);
     const {
+      StartPage,
+      datasourceInstance,
+      datasourceMissing,
       initialized,
-      exploreId,
+      showingStartPage,
+      queryKeys,
+      urlState,
+      update,
+      isLive,
+      supportedModes,
+      mode,
+      graphResult,
+      loading,
+      showingGraph,
+      showingTable,
+      absoluteRange,
+      queryResponse,
+    } = item;
+
+    const { datasource, queries, range: urlRange, mode: urlMode, ui, originPanelId } = (urlState ||
+      {}) as ExploreUrlState;
+    const initialDatasource = datasource || store.get(lastUsedDatasourceKeyForOrgId(state.user.orgId));
+    const initialQueries: DataQuery[] = ensureQueriesMemoized(queries);
+    const initialRange = urlRange ? getTimeRangeFromUrlMemoized(urlRange, timeZone).raw : DEFAULT_RANGE;
+
+    let newMode: ExploreMode;
+
+    if (supportedModes.length) {
+      const urlModeIsValid = supportedModes.includes(urlMode);
+      const modeStateIsValid = supportedModes.includes(mode);
+
+      if (modeStateIsValid) {
+        newMode = mode;
+      } else if (urlModeIsValid) {
+        newMode = urlMode;
+      } else {
+        newMode = supportedModes[0];
+      }
+    } else {
+      newMode = [ExploreMode.Metrics, ExploreMode.Logs].includes(urlMode) ? urlMode : null;
+    }
+
+    const initialUI = ui || DEFAULT_UI_STATE;
+
+    return {
+      StartPage,
+      datasourceInstance,
+      datasourceMissing,
+      initialized,
+      showingStartPage,
+      split,
+      queryKeys,
+      update,
       initialDatasource,
       initialQueries,
       initialRange,
-      mode,
+      mode: newMode,
       initialUI,
+      isLive,
+      graphResult,
+      loading,
+      showingGraph,
+      showingTable,
+      absoluteRange,
+      queryResponse,
       originPanelId,
-    } = this.props;
+      syncedTimes,
+    };
+  }
+
+  actionsToDispatch(): ReduxActions {
+    return {
+      changeSize,
+      initializeExplore,
+      modifyQueries,
+      refreshExplore: refreshExplore,
+      scanStart,
+      scanStopAction,
+      setQueries,
+      updateTimeRange,
+      toggleGraph,
+    };
+  }
+
+  componentDidMount() {
+    const { exploreId } = this.props;
+    const { initialized, initialDatasource, initialQueries, initialRange, mode, initialUI, originPanelId } = this.state;
     const width = this.el ? this.el.offsetWidth : 0;
 
     // initialize the whole explore first time we mount and if browser history contains a change in datasource
     if (!initialized) {
-      this.props.initializeExplore(
+      this.actions.initializeExplore(
```

Comment:
The component will have an `actions` prop that will contain all `Redux bounded Thunks or ActionCreators`